### PR TITLE
Sub-organizations smart search

### PR DESCRIPTION
**Pull request by richardolsson**
## Description
This PR implements support for sub-organization scoping and data in Smart Searches. With the features introduced in this PR, it becomes possible to search for people based on what happened in sub-organizations, e.g. having filled out a survey belonging to a sub-org or having attended events organized by a sub-org.

This entails two main things:

1. It needs to be possible to scope filters to organizations, i.e. configuring what organizations a filter should be searching within
2. Data belonging to sub-organizations needs to be accessible where applicable when configuring a filter, e.g. when configuring a survey filter the surveys of sub-organizations need to be available to pick from.

This PR fully implements 1, and partially solves 2 (see notes to reviewer).

### Editing a filter with organization scoping
Screenshot 2024-04-22 at 12 14 03

### Selecting organization scope
Screenshot 2024-04-22 at 12 13 10

### Configuring with sub-organization content
Screenshot 2024-04-22 at 12 14 14

### Scope displayed in overview
Screenshot 2024-04-22 at 12 14 40

## Changes
* Adds an organization scope widget to all (applicable) filter configs
* Adds logic to display the scoping in the Smart Search overview (with the Sankey diagram)
* Implements pattern for loading data belonging to sub-organizations (see notes to reviewer)
* Adds new `StyledGroupedSelect` widget to display data grouped by organization

## Notes to reviewer
This PR only introduces the sub-org data loading for a single type of data, in a single filter, namely campaigns/projects in the `campaign_participation` filter. I want to test the pattern and have it reviewed before I proceed to implement the same logic for all other filters and data types.

## Related issues
I first thought that this was undocumented, but it's actually related to (but does not yet fully fix) #1858

### CODE REVIEW

**Review by ziggabyte — CHANGES_REQUESTED** [review]
This is a very cool feature, and also cool that it came about so quickly. Surely this will mean a significant improvement to the smart search world! 

I am marking this as requesting changes, because I have some feedback on naming + I did encounter two strange behaviours when playing with it: 

1. Sub-org search is not working as expected when searching custom fields (which feels reasonable, since those orgs might not have the same custom fields). Is it maybe better to not have the sub-orgs selector in that filter? 

https://github.com/zetkin/app.zetkin.org/assets/58265097/a7d15c07-f70c-4b93-af41-e4093139782f

2. The whole smart search returns 0 when adding a filter where data from a sub-org returns 0 (i think?). What made me notice it is that the sankey diagram disappears, but the number on the screen tells you that the whole smart search returns 0 people after adding that filter. 


https://github.com/zetkin/app.zetkin.org/assets/58265097/cb5827d4-774e-44bd-a57b-fabce6f43263

File: src/core/caching/shouldLoad.ts [hunk 1]
```diff
@@ -2,6 +2,22 @@ import { RemoteItem, RemoteList } from 'utils/storeUtils';
...
 const DEFAULT_TTL = 5 * 60 * 1000; // 5 minutes
 
+type ObjThatNeedsLoading =
+  | Record<number | string, RemoteList<unknown> | RemoteItem<unknown>>
+  | RemoteItem<unknown>
+  | RemoteList<unknown>
+  | undefined;
+
+function isMap(
+  thing: ObjThatNeedsLoading
```
[the remark this answers is not in the document]
Comment (richardolsson) [reply src/core/caching/shouldLoad.ts:12]: The name "Map" refers to the conceptual type of data container which in Javascript is realized technically using either the `Object` type or the `Map` type. It's a mapping between two values, in this context the keys are IDs and the values are list of objects related to that ID. For example, a list of campaigns related to an organization.

You can read more about maps (or associative arrays) on Wikipedia: https://en.wikipedia.org/w/index.php?title=Map_(computer_science)

The reason for the name `thing` (which was already used elsewhere in this file) is that at the point where the variable is declared, we don't yet know what it is, just that it is some thing. This function will return whether or not that thing is a map.

**Comment by richardolsson** [comment]
We found a bug, which is that it's possible to pick inactive organizations when selecting "specific organizations". We should filter out the inactive organizations there.

**Review by ziggabyte — APPROVED** [review]
The changes were changed - things look good - let's merge! :)

**Pull request merged**